File: backend/dao/repositories/security.py
from __future__ import annotations

import json
from typing import Any
from uuid import UUID

from pkg.apis.security.models import DetectionResult, SecurityResult
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from dao.models.security import DetectorExecution, SecurityChunkResult, SecurityScan


class SecurityScanRepository:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def create_from_result(
        self,
        result: SecurityResult,
        *,
        prompt_hash: str,
        prompt_length: int,
        prompt_text: str | None,
        user_id: UUID | None = None,
        worker_id: str | None = None,
    ) -> SecurityScan:
        metadata = result.metadata
        scan = SecurityScan(
            request_id=result.request_id,
            user_id=user_id,
            decision=result.decision.value,
            score=result.score,
            threshold=metadata.get("threshold"),
            detector=str(metadata.get("detector", "unknown")),
            detector_version=str(metadata.get("detector_version", "unknown")),
            policy=result.policy,
            chunks_scanned=result.chunks_scanned,
            chunking_strategy=str(metadata.get("chunker", "fixed")),
            latency_ms=result.total_latency_ms,
            prompt_hash=prompt_hash,
            prompt_length=prompt_length,
            prompt_text=prompt_text,
            model_target=metadata.get("model"),
            worker_id=worker_id,
            source=metadata.get("source"),
            metadata_json=json.dumps(metadata, default=str),
        )
        self.session.add(scan)
        await self.session.flush()

        for index, detection in enumerate(result.detector_results):
            self.session.add(
                SecurityChunkResult(
                    scan_id=scan.id,
                    chunk_index=index,
                    detector=detection.detector,
                    score=detection.score,
                    is_injection=detection.is_injection,
                    latency_ms=detection.latency_ms,
                )
            )
            self.session.add(
                DetectorExecution(
                    scan_id=scan.id,
                    detector=detection.detector,
                    detector_version=str(detection.metadata.get("version", "unknown")),
                    threshold=detection.threshold,
                    score=detection.score,
                    is_injection=detection.is_injection,
                    latency_ms=detection.latency_ms,
                    metadata_json=json.dumps(detection.metadata, default=str),
                )
            )
        await self.session.flush()
        return scan

    async def get(self, scan_id: UUID) -> SecurityScan | None:
        result = await self.session.execute(select(SecurityScan).where(SecurityScan.id == scan_id))
        return result.scalar_one_or_none()

    async def list_scans(self, *, limit: int = 50, offset: int = 0) -> list[SecurityScan]:
        result = await self.session.execute(
            select(SecurityScan)
            .order_by(SecurityScan.created_at.desc())
            .limit(limit)
            .offset(offset)
        )
        return list(result.scalars().all())

    async def chunk_results(self, scan_id: UUID) -> list[SecurityChunkResult]:
        result = await self.session.execute(
            select(SecurityChunkResult)
            .where(SecurityChunkResult.scan_id == scan_id)
            .order_by(SecurityChunkResult.chunk_index)
        )
        return list(result.scalars().all())

    async def detector_executions(self, scan_id: UUID) -> list[DetectorExecution]:
        result = await self.session.execute(
            select(DetectorExecution).where(DetectorExecution.scan_id == scan_id)
        )
        return list(result.scalars().all())

    async def summary_counts(self) -> dict[str, Any]:
        total = await self.session.scalar(select(func.count(SecurityScan.id))) or 0
        allowed = await self.session.scalar(
            select(func.count(SecurityScan.id)).where(SecurityScan.decision == "ALLOW")
        ) or 0
        blocked = await self.session.scalar(
            select(func.count(SecurityScan.id)).where(SecurityScan.decision == "BLOCK")
        ) or 0
        review = await self.session.scalar(
            select(func.count(SecurityScan.id)).where(SecurityScan.decision == "REVIEW")
        ) or 0
        avg_latency = await self.session.scalar(select(func.avg(SecurityScan.latency_ms))) or 0.0
        return {
            "total_scans": int(total),
            "allowed": int(allowed),
            "blocked": int(blocked),
            "review": int(review),
            "avg_detection_latency_ms": float(avg_latency),
        }

    def detection_results_from_chunks(
        self, chunks: list[SecurityChunkResult]
    ) -> list[DetectionResult]:
        return [
            DetectionResult(
                detector=chunk.detector,
                score=chunk.score,
                is_injection=chunk.is_injection,
                latency_ms=chunk.latency_ms,
            )
            for chunk in chunks
        ]

File: backend/dao/store/security.go
package store

import (
	"context"
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/spider/spider/pkg/apis"
)

type SecurityScan struct {
	ID               uuid.UUID
	RequestID        uuid.UUID
	UserID           *uuid.UUID
	Decision         string
	Score            float64
	Threshold        *float64
	Detector         string
	DetectorVersion  string
	Policy           string
	ChunksScanned    int
	ChunkingStrategy string
	LatencyMs        float64
	PromptHash       string
	PromptLength     int
	PromptText       *string
	ModelTarget      *string
	WorkerID         *string
	Source           *string
	MetadataJSON     string
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

type SecurityRepo struct {
	pool *pgxpool.Pool
}

func NewSecurityRepo(pool *pgxpool.Pool) *SecurityRepo {
	return &SecurityRepo{pool: pool}
}

func (r *SecurityRepo) CreateFromResult(ctx context.Context, result apis.SecurityResult, promptHash string, promptLength int, promptText *string, userID *uuid.UUID, workerID *string) (*SecurityScan, error) {
	meta := result.Metadata
	if meta == nil {
		meta = map[string]interface{}{}
	}
	metaJSON, _ := json.Marshal(meta)
	now := time.Now().UTC()
	id := uuid.New()

	var threshold *float64
	if v, ok := meta["threshold"].(float64); ok {
		threshold = &v
	}
	detector := strMeta(meta, "detector", "unknown")
	detectorVersion := strMeta(meta, "detector_version", "unknown")
	chunker := strMeta(meta, "chunker", "fixed")
	var modelTarget, source *string
	if v, ok := meta["model"].(string); ok {
		modelTarget = &v
	}
	if v, ok := meta["source"].(string); ok {
		source = &v
	}

	reqID, _ := uuid.Parse(result.RequestID)

	_, err := r.pool.Exec(ctx, `
		INSERT INTO security_scans (id, request_id, user_id, decision, score, threshold, detector,
			detector_version, policy, chunks_scanned, chunking_strategy, latency_ms, prompt_hash,
			prompt_length, prompt_text, model_target, worker_id, source, metadata_json, created_at, updated_at)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21)`,
		id, reqID, userID, string(result.Decision), result.Score, threshold, detector,
		detectorVersion, result.Policy, result.ChunksScanned, chunker, result.TotalLatencyMs,
		promptHash, promptLength, promptText, modelTarget, workerID, source, string(metaJSON), now, now)
	if err != nil {
		return nil, err
	}

	for index, detection := range result.DetectorResults {
		dmeta, _ := json.Marshal(detection.Metadata)
		_, err = r.pool.Exec(ctx, `
			INSERT INTO security_chunk_results (id, scan_id, chunk_index, detector, score, is_injection, latency_ms, created_at, updated_at)
			VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)`,
			uuid.New(), id, index, detection.Detector, detection.Score, detection.IsInjection,
			detection.LatencyMs, now, now)
		if err != nil {
			return nil, err
		}
		version := strMeta(detection.Metadata, "version", "unknown")
		_, err = r.pool.Exec(ctx, `
			INSERT INTO detector_executions (id, scan_id, detector, detector_version, threshold, score,
				is_injection, latency_ms, metadata_json, created_at, updated_at)
			VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11)`,
			uuid.New(), id, detection.Detector, version, detection.Threshold, detection.Score,
			detection.IsInjection, detection.LatencyMs, string(dmeta), now, now)
		if err != nil {
			return nil, err
		}
	}

	return &SecurityScan{
		ID: id, RequestID: reqID, UserID: userID, Decision: string(result.Decision),
		Score: result.Score, Threshold: threshold, Detector: detector,
		DetectorVersion: detectorVersion, Policy: result.Policy,
		ChunksScanned: result.ChunksScanned, ChunkingStrategy: chunker,
		LatencyMs: result.TotalLatencyMs, PromptHash: promptHash, PromptLength: promptLength,
		PromptText: promptText, ModelTarget: modelTarget, WorkerID: workerID, Source: source,
		MetadataJSON: string(metaJSON), CreatedAt: now, UpdatedAt: now,
	}, nil
}

func (r *SecurityRepo) Get(ctx context.Context, scanID uuid.UUID) (*SecurityScan, error) {
	row := r.pool.QueryRow(ctx, `
		SELECT id, request_id, user_id, decision, score, threshold, detector, detector_version,
			policy, chunks_scanned, chunking_strategy, latency_ms, prompt_hash, prompt_length,
			prompt_text, model_target, worker_id, source, metadata_json, created_at, updated_at
		FROM security_scans WHERE id = $1`, scanID)
	return scanSecurityRow(row)
}

func (r *SecurityRepo) ListScans(ctx context.Context, limit, offset int) ([]SecurityScan, error) {
	rows, err := r.pool.Query(ctx, `
		SELECT id, request_id, user_id, decision, score, threshold, detector, detector_version,
			policy, chunks_scanned, chunking_strategy, latency_ms, prompt_hash, prompt_length,
			prompt_text, model_target, worker_id, source, metadata_json, created_at, updated_at
		FROM security_scans ORDER BY created_at DESC LIMIT $1 OFFSET $2`, limit, offset)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []SecurityScan
	for rows.Next() {
		s, err := scanSecurityRows(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *s)
	}
	return out, rows.Err()
}

type ChunkResult struct {
	ChunkIndex  int
	Detector    string
	Score       float64
	IsInjection bool
	LatencyMs   float64
}

type DetectorExecution struct {
	Detector        string
	DetectorVersion string
	Threshold       *float64
	Score           float64
	IsInjection     bool
	LatencyMs       float64
	MetadataJSON    string
}

func (r *SecurityRepo) ChunkResults(ctx context.Context, scanID uuid.UUID) ([]ChunkResult, error) {
	rows, err := r.pool.Query(ctx, `
		SELECT chunk_index, detector, score, is_injection, latency_ms
		FROM security_chunk_results WHERE scan_id = $1 ORDER BY chunk_index`, scanID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []ChunkResult
	for rows.Next() {
		var c ChunkResult
		if err := rows.Scan(&c.ChunkIndex, &c.Detector, &c.Score, &c.IsInjection, &c.LatencyMs); err != nil {
			return nil, err
		}
		out = append(out, c)
	}
	return out, rows.Err()
}

func (r *SecurityRepo) DetectorExecutions(ctx context.Context, scanID uuid.UUID) ([]DetectorExecution, error) {
	rows, err := r.pool.Query(ctx, `
		SELECT detector, detector_version, threshold, score, is_injection, latency_ms, metadata_json
		FROM detector_executions WHERE scan_id = $1`, scanID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []DetectorExecution
	for rows.Next() {
		var d DetectorExecution
		if err := rows.Scan(&d.Detector, &d.DetectorVersion, &d.Threshold, &d.Score, &d.IsInjection, &d.LatencyMs, &d.MetadataJSON); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

func (r *SecurityRepo) RecordScanMetric(ctx context.Context, decision string, latencyMs float64) error {
	_, err := r.pool.Exec(ctx, `
		INSERT INTO scan_metrics_counters (decision, count, total_latency_ms)
		VALUES ($1, 1, $2)
		ON CONFLICT (decision) DO UPDATE
		SET count = scan_metrics_counters.count + 1,
		    total_latency_ms = scan_metrics_counters.total_latency_ms + EXCLUDED.total_latency_ms
	`, decision, latencyMs)
	return err
}

func (r *SecurityRepo) SummaryCounts(ctx context.Context) (map[string]interface{}, error) {
	// Try aggregate counters table first for zero-loss telemetry
	rows, err := r.pool.Query(ctx, `SELECT decision, count, total_latency_ms FROM scan_metrics_counters`)
	if err == nil {
		defer rows.Close()
		var total, allowed, blocked, review int64
		var totalLatency float64
		hasData := false
		for rows.Next() {
			var dec string
			var cnt int64
			var lat float64
			if err := rows.Scan(&dec, &cnt, &lat); err == nil {
				hasData = true
				total += cnt
				totalLatency += lat
				switch dec {
				case "ALLOW":
					allowed += cnt
				case "BLOCK":
					blocked += cnt
				case "REVIEW":
					review += cnt
				}
			}
		}
		if hasData {
			avgLatency := 0.0
			if total > 0 {
				avgLatency = totalLatency / float64(total)
			}
			return map[string]interface{}{
				"total_scans":              int(total),
				"allowed":                  int(allowed),
				"blocked":                  int(blocked),
				"review":                   int(review),
				"avg_detection_latency_ms": avgLatency,
			}, nil
		}
	}

	// Fallback to querying security_scans table if counters table is not yet populated
	var total, allowed, blocked, review int
	var avgLatency float64
	err = r.pool.QueryRow(ctx, `SELECT COUNT(*) FROM security_scans`).Scan(&total)
	if err != nil {
		return nil, err
	}
	_ = r.pool.QueryRow(ctx, `SELECT COUNT(*) FROM security_scans WHERE decision='ALLOW'`).Scan(&allowed)
	_ = r.pool.QueryRow(ctx, `SELECT COUNT(*) FROM security_scans WHERE decision='BLOCK'`).Scan(&blocked)
	_ = r.pool.QueryRow(ctx, `SELECT COUNT(*) FROM security_scans WHERE decision='REVIEW'`).Scan(&review)
	_ = r.pool.QueryRow(ctx, `SELECT COALESCE(AVG(latency_ms), 0) FROM security_scans`).Scan(&avgLatency)
	return map[string]interface{}{
		"total_scans": total, "allowed": allowed, "blocked": blocked, "review": review,
		"avg_detection_latency_ms": avgLatency,
	}, nil
}

func strMeta(m map[string]interface{}, key, fallback string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return fallback
}

type scannable interface {
	Scan(dest ...interface{}) error
}

func scanSecurityRow(row scannable) (*SecurityScan, error) {
	var s SecurityScan
	err := row.Scan(&s.ID, &s.RequestID, &s.UserID, &s.Decision, &s.Score, &s.Threshold,
		&s.Detector, &s.DetectorVersion, &s.Policy, &s.ChunksScanned, &s.ChunkingStrategy,
		&s.LatencyMs, &s.PromptHash, &s.PromptLength, &s.PromptText, &s.ModelTarget,
		&s.WorkerID, &s.Source, &s.MetadataJSON, &s.CreatedAt, &s.UpdatedAt)
	if err != nil {
		return nil, err
	}
	return &s, nil
}

func scanSecurityRows(rows scannable) (*SecurityScan, error) {
	return scanSecurityRow(rows)
}
